src/tactic/transport.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner, Simon Hudon, Scott Morrison
+-/
+import tactic.equiv_rw
+
+/-!
+## The `transport` tactic
+
+`transport` attempts to move an `s : S α` expression across an equivalence `e : α ≃ β` to solve
+a goal of the form `S β`, by building the new object field by field, taking each field of `s`
+and rewriting it along `e` using the `equiv_rw` tactic.
+
+We try to ensure good definitional properties, so that, for example, when we transport a `monoid α`
+to a `monoid β`, the new multiplication is definitionally `λ x y, e (e.symm a * e.symm b)`.
+-/
+
+namespace tactic
+open tactic.interactive
+
+mk_simp_attribute transport_simps
+"The simpset `transport_simps` is used by the tactic `transport`
+to simplify certain expressions involving application of equivalences,
+and trivial `eq.rec` or `ep.mpr` conversions.
+It's probably best not to adjust it without understanding the algorithm used by `transport`."
+
+attribute [transport_simps]
+  eq_rec_constant
+  eq_mpr_rfl
+  equiv.to_fun_as_coe
+  equiv.arrow_congr'_apply
+  equiv.symm_apply_apply
+  -- we use `apply_eq_iff_eq_symm_apply` rather than `apply_eq_iff_eq`,
+  -- as many axioms have a constant on the right-hand-side
+  equiv.apply_eq_iff_eq_symm_apply
+
+/--
+Given `s : S α` for some structure `S` depending on a type `α`,
+and an equivalence `e : α ≃ β`,
+try to produce an `S β`,
+by transporting data and axioms across `e` using `equiv_rw`.
+-/
+@[nolint unused_arguments] -- At present we don't actually use `s`; it's inferred in the `mk_app` call.
+meta def transport (s e : expr) : tactic unit :=
+do
+  (_, α, β) ← infer_type e >>= relation_lhs_rhs <|>
+    fail format!"second argument to `transport` was not an equivalence-type relation",
+  -- We explode the goal into individual fields using `refine_struct`.
+  -- Later we'll want to also consider falling back to `fconstructor`,
+  -- but for now this suffices.
+  seq `[refine_struct { .. }]
+  -- We now deal with each field sequentially.
+  -- Since we may pass goals through to the user if the heuristics here fail,
+  -- we wrap everything in `propagate_tags`.
+  (propagate_tags (try (do
+    -- Look up the name of the current field being processed
+    f ← get_current_field,
+    -- Note the value in the original structure.
+    -- (This `mk_mapp` call with second argument inferred only works for typeclasses!)
+    mk_mapp f [α, none] >>= note f none,
```

Comment:
I think this is the biggest todo here because it fails in multiple ways:

 - It only works for type classes and not for structures, contrary to the docstring.
 - It doesn't work for type classes with extra parameters (such as `[monoid α]`, etc.)

Unfortunately `mk_mapp f [α, s]` doesn't work due to inheritance.

I think the proper and reliable way to implement this is to eschew `refine_struct` and make `transport s e` a recursive function (that transports `s` along `e` to the current target).  That is, if the type of `s` is a structure then make a constructor (potentially using `fconstructor`) and recursively call `transport proj_s e` for each field with the appropriate projection of `s`.  Otherwise run the current logic for propositions and data.

Reply:
I will investigate this; it seems a good idea.

I couldn't quite tell from the combination of "bors d+" and this comment whether you think it's okay to make this change in a future PR? If so, I'll add comments here explaining the future plan, then merge.

Reply:
I think this can go into a future PR.